### DISCUSSION TITLE
[Done] non-SMTP transport to email package?

### DISCUSSION THREAD

**Discussion by xet7** [meteor/meteor#11547] [Feature Requests]
Hi,
for nodemailer service, it looks like email package only has SMTP transport?
- https://github.com/meteor/meteor/blob/devel/packages/email/email.js#L26
- https://github.com/meteor/meteor/blob/devel/packages/email/email.js#L156

Or is it possible to select another `service`? So if protocol is not `smtp`, for example

a) use protocol with URL like `Outlook365://username:password@...` ? Nodemailser `service`s are listed here: https://nodemailer.com/smtp/well-known/ .

b) is it possible to add `service` to `smtp` url like `smtp://....../?service=Outlook365` ?

For example, with this nodemailer code I can send email using O365:

```
var nodemailer = require('nodemailer');

let transporter = nodemailer.createTransport({
  service: "Outlook365",
  auth: {
    user: 'firstname.lastname@hotmail.com',
    pass: 'topsecret'
  },
})

let info = transporter.sendMail({
  from: 'Support Team <firstname.lastname@hotmail.com>',
  to: 'customer@example.com',
  subject: 'Hi',
  text: 'Hi! Reminder: Meeting tomorrow.',
  html: '<h1>Hi! Reminder: Meeting tomorrow.</h1>'
})
```

Well, that email package is small amount of code, so I maybe could fork it, but I have already done so much forking I kind of have too big fork factory already ;)

**Comment by xet7** [comment]
@StorytellerCZ commented today at Meteor Slack: "I plan to start next week working on a new version of the e-mail package. I will look into this."

When I tested sending with O365 using this `service: "Outlook365"`, it did work.

But using [SMTP with command line](https://github.com/wekan/wekan/wiki/Troubleshooting-Mail#example-aws-ses-sending-email-with-command-line) for `smtp.office365.com:587` I got these errors both with STARTTLS and TLS, so using SMTP did not work at all. There could be some predefined correct settings and behaviours that are already included in working  `service: "Outlook365"`.

```
535 5.7.3 Authentication unsuccessful [HE1PR0202CA0021.eurprd02.prod.outlook.com]
530 5.7.57 Client not authenticated to send mail. Error: 535 5.7.3 Authentication unsuccessful [HE1PR0202CA0021.eurprd02.prod.outlook.com]
read:errno=0
```

**Reply by StorytellerCZ** [reply]
Interesting.

Anyhow I'm going add this to my list for the next version of e-mail, this currently includes:
- [x] modernization
- [x] Send out email with custom API
- [x] NEW: implement nodemailer's known hosts
- [ ] NEW: New way to set SMTP connection string

Now for the two new features we need to decide on the API.

## New way to set SMTP connection string
I think this is the start of everything else here. Currently the SMTP connection string is set via environment variable (`MAIL_URL`). While maintaining backward compatibility I think we should add an option to set this in the app settings say under `packages.email.mailUrl` (or something along those lines). The only question left is which url should take preference.

## Implement nodemailer's known hosts
Along with the previous task we could add an additional setting where a known service could be specified instead. Say `packages.email.service` which would be an object like:
```js
{
  name: '1on1',
  user: 'mdg',
  password: 'password'
}
```
Then if a service is defined then that will be recognized before it even gets to the url check and return a mailer based on that service.

## Send out email with custom API
While this is currently possible with `sendHook`, I would like to improve upon this. I see two possible approached here.
First add a new variable like `Email.customSend = (options) => {}` which if set by the app would take precedent over SMTP send in production, while using the dev send in development.
Second more complex approach, or the next step, would be to make the current `email` package a core and then you would need to add a new package with the provider you want to use in the same way as it is with the accounts packages.


Thoughts?

**Reply by StorytellerCZ** [reply]
I will also look into implementing other `nodemailer` functionality.

UPDATE:
So I thought the following functionality could be added:
- [ ] Add support for proxies (just the first example from the [docs](https://nodemailer.com/smtp/proxies/)), just another field into the app config
- [ ] [DSN](https://nodemailer.com/smtp/dsn/)

**Reply by xet7** [reply]
a) If it would be possible for exising MAIL_URL to work like these, then I would not need any changes to current Wekan settings, only update mail package, I would prefer this option:

- `smtp://username:password@...`
- `smtps://username:password@...`
- `Outlook365://username:password@...`

In Meteor email package code, if protocol is not `smtp` or `smtps`, then it would try to use that protocol as `service` for transport.

b) But if it would be `smtp://....../?service=Outlook365`, it's more confusing, because is it smtp or other service? In worst case, I would need to handle parsing `MAIL_URL` options myself in special Wekan code, if those options are not handled in Meteor email package itself. Previously there used to be separate web UI at Wekan Admin Panel to set each email option, but it was so buggy and did not always work, so now I currently only have `MAIL_URL` setting. If those settings are not at same `MAIL_URL`, then I would need to add separate environment variable options for various settings, adding more complexity.

**Reply by xet7** [reply]
Although, in any case, I would prefer most dependencies being in upstream packages, and being updated with `meteor update`, because manually merging changes to each forked dependency is so much extra work.

**Reply by StorytellerCZ** [reply]
I prefer the first option here as it is easier to implement with less things that can go wrong.

**Comment by filipenevola** [comment]
@StorytellerCZ are your changes in a branch?

It would be nice to already review the changes that are already done.

**Reply by xet7** [reply]
I will try to make some temporary fork to Wekan with only that `service` to `Outlook365://` fix, so Wekan users can get their email sending working, they are daily asking me when fix will be available.

**Reply by StorytellerCZ** [reply]
This is the remaining task before I have a PR, I'm almost done with it. See: https://github.com/meteor/meteor/pull/11554

**Comment by StorytellerCZ** [comment]
This has been done and released in Meteor 2.4 as part of email@2.2
